src/runtime/testing/slot-test-observer.ts

```diff
@@ -0,0 +1,286 @@
+ *           .expectRenderSlot('MyParticle1', 'mySlot1');
+ *           .expectRenderSlot('MyParticle1', 'mySlot1', {times: 2})
+ *           .expectRenderSlot('MyParticle2', 'mySlot2', {verify: (content) => !!content.myParam})
+ *           .expectRenderSlot('MyOptionalParticle', 'myOptionalSlot', {isOptional: true})
+ *   await observer.expectationsCompleted();
+ */
+export class SlotTestObserver extends AbstractSlotObserver {
+  private expectQueue;
+  onExpectationsComplete;
+  strict: boolean;
+  logging: boolean;
+  debugMessages;
+  pec;
+
+  /**
+   * |options| may contain:
+   * - strict: whether unexpected render slot requests cause an assert or a warning log (default: true)
+   */
+  constructor(options: SlotTestObserverOptions = {}) {
+    super();
+    this.expectQueue = [];
+    this.onExpectationsComplete = () => undefined;
+    this.strict = options.strict != undefined ? options.strict : true;
+    this.logging = Boolean(options.logging);
+    this.debugMessages = [];
+    // Clear all cached templates
+    HeadlessSlotDomConsumer.clearCache();
+  }
+
+  /**
+   * Reinitializes expectations queue.
+   */
+  newExpectations(name?: string) {
+    this.expectQueue = [];
+    if (!this.strict) {
+      this.ignoreUnexpectedRender();
+    }
+    this.debugMessages.push({name: name || `debug${Object.keys(this.debugMessages).length}`, messages: []});
+    return this;
+  }
+
+  /**
+   * Allows ignoring unexpected render slot requests.
+   */
+  ignoreUnexpectedRender() {
+    this.expectQueue.push({type: 'render', ignoreUnexpected: true, isOptional: true,
+                           toString: () => `render: ignoreUnexpected optional`});
+    return this;
+  }
+
+  /**
+   * Returns true, if the number of items in content's model is equal to the given number.
+   */
+  expectContentItemsNumber(num: number, content) {
+    assert(content.model, `Content doesn't have model`);
+    assert(content.model.items, `Content model doesn't have items (${num} expected}`);
+    assert(content.model.items.length <= num, `Too many items (${content.model.items.length}), while only ${num} were expected.`);
+    return content.model.items.length === num;
+  }
+
+  /**
+   * Adds a rendering expectation for the given particle and slot names, where options may contain:
+   * times: number of time the rendering request will occur
+   * contentTypes: the types appearing in the rendering content
+   * isOptional: whether this expectation is optional (default: false)
+   * hostedParticle: for transformation particles, the name of the hosted particle
+   * verify: an additional optional handler that determines whether the incoming render request satisfies the expectation
+   */
+  expectRenderSlot(particleName, slotName, options?) {
+    options = options || {};
+    const times = options.times || 1;
+    for (let i = 0; i < times; ++i) {
+      this.addRenderExpectation({
+        particleName,
+        slotName,
+        contentTypes: options.contentTypes,
```

Comment:
weren't `contentTypes` deprecated (or maybe this is in a future PR)?

Reply:
Yes, deprecated, I should remove it, thanks.